src/Tools/BuildValidator/LocalReferenceResolver.cs

```diff
@@ -0,0 +1,126 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.Extensions.Logging;
+
+namespace BuildValidator
+{
+    /// <summary>
+    /// Resolves references for a package by looking in local nuget and artifact
+    /// directories for Roslyn
+    /// </summary>
+    internal class LocalReferenceResolver : IMetadataReferenceResolver
+    {
+        private readonly Dictionary<Guid, string> _cache = new Dictionary<Guid, string>();
+        private readonly HashSet<DirectoryInfo> _indexDirectories = new HashSet<DirectoryInfo>();
+        private readonly ILogger _logger;
+
+        public LocalReferenceResolver(ILoggerFactory loggerFactory)
+        {
+            _logger = loggerFactory.CreateLogger<LocalReferenceResolver>();
+            _indexDirectories.Add(GetArtifactsDirectory());
+            _indexDirectories.Add(GetNugetCacheDirectory());
+        }
+
+        public static DirectoryInfo GetNugetCacheDirectory()
+        {
+            var nugetPackageDirectory = Environment.GetEnvironmentVariable("NUGET_PACKAGES");
+            if (nugetPackageDirectory is null)
+            {
+                nugetPackageDirectory = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".nuget");
+            }
+
+            return new DirectoryInfo(nugetPackageDirectory);
+        }
+
+        public Task<MetadataReference> ResolveReferenceAsync(MetadataReferenceInfo referenceInfo)
+        {
+            var path = Search(referenceInfo.Mvid, referenceInfo.Name);
+            return Task.FromResult<MetadataReference>(MetadataReference.CreateFromFile(path));
+        }
+
+        public Task<ImmutableArray<MetadataReference>> ResolveReferencesAsync(IEnumerable<MetadataReferenceInfo> references)
+        {
+            var referenceArray = references.ToImmutableArray();
+            CacheNames(referenceArray);
+
+            var files = referenceArray.Select(r => _cache[r.Mvid]);
+
+            var metadataReferences = files.Select(f => MetadataReference.CreateFromFile(f)).Cast<MetadataReference>().ToImmutableArray();
```

Comment:
The creation of the `MetadataReference` should apply the extern aliases in the associated `MetadataReferenceInfo` correct?